airbyte-cdk/python/airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py

```diff
@@ -350,7 +360,24 @@ def _create_nested_component(self, model, model_field: str, model_value: Any, co
         model_type = self.TYPE_NAME_TO_MODEL.get(type_name, None)
         if model_type:
             parsed_model = model_type.parse_obj(model_value)
-            return self._create_component_from_model(model=parsed_model, config=config)
+            try:
+                # To improve usability of the language, certain fields are shared between components. One example is the DefaultPaginator
+                # referencing the HttpRequester url_base while constructing a SimpleRetriever. However, custom components don't support
+                # this behavior because they are created generically in create_custom_component(). This block allows developers to
+                # specify extra arguments in $parameters that are needed by a component and could not be shared.
```

Comment:
I think you bring up a good point that this also services a use case where parent custom component can't pass a child a special value. I can modify the comment to include this. 

But I think we also need to retain the context that this supports a more broad use case. The problem we're observing here is the field that we want to use for child is not actually stored on the parent custom retriever itself. In this case, the `url_base` is on the HttpRequester, which is "adjacent" to the DefaultPaginator. So this goes beyond just auto-propagating and prior to this change, we simply don't know what fields we need to propagate back out.

So going back to the general question in your review, while I agree it might be simpler to auto-propagate out and help avoid specifying this extra `$parameter`, it wouldn't solve for the original case because `url_base` doesn't exist on the custom retriever so it wouldn't be passed to DefaultPaginator.

Let me know if this all makes sense, but happy to dig a little deeper, this is kind of a weird edge that I don't think any of us had considered until a few days ago. Sometimes custom components really do be like that

Reply:
> So just to restate the more general problem - it looks like in the "normal" case, we enforce an order - the requester is created before the paginator. This suggests a kind of dependency, even though like you pointed out, the components themselves are defined adjacently.

We don't quite need an order exactly (at least not in this case) because we are relying on the models instead of the declarative components being built. For example in our example, the retriever pulls the `url_base` from the http requester Pydantic model (as opposed to the resulting component), so you could in theory build the paginator first since we're relying on the fully parsed models.

So in a way our problem that we're faced with is at least a little simpler, we just need some sort of record that can tell us for a given component, what dependent fields it has and where it can find it. Although maybe there are some questions about how and in what manner we store this. I agree that the unification of these flows is the end goal we want to reach

Reply:
Okay cool, that makes sense to me!